sw/cheri/common/platform-pinmux.hh.tpl

```diff
@@ -1,195 +1,201 @@
+	  return ${prev_num_options};
 % endif
-      default:
-        return 2;
-    }
+	default:
+	  return 2;
   }
+}
 
-  /**
-   * A helper function that returns the number of input pins that can be
-   * selected from in the pin multiplexer for a given block input. This will
-   * always be at least 2, as option 0 represents 'OFF' i.e. no connection,
-   * and option 1 represents the default connection.
-   *
-   * @param block_input The block input to query.
-   * @returns The number of selections available for that block input.
-   *
-   * The meanings of these selections can be found in the documentation:
-   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
-   */
-  static constexpr uint8_t block_input_options(BlockInput block_input) {
-    switch (block_input) {
+/**
+ * Returns the number of sources available for a block sink (block input).
+ *
+ * @param block_sink The block sink to query.
+ * @returns The number of sources available for the given sink.
+ */
+static constexpr uint8_t sources_number(BlockSink block_sink) {
+  switch (block_sink) {
 <%
-    prev_num_options = 0
+  prev_num_options = 0
 %>
 % for (block_io, _, num_options) in filter(lambda block: block[2] > 2, sorted(output_block_ios, key=lambda block: block[2], reverse=True)):
-    % if prev_num_options != 0 and prev_num_options != num_options:
-        return ${prev_num_options};
-    % endif
-    <%
-        prev_num_options = num_options 
-    %>      case BlockInput::${block_io.doc_name.replace("[","_").replace(".","_").replace("]","")}:
+  % if prev_num_options != 0 and prev_num_options != num_options:
+	  return ${prev_num_options};
+  % endif
+  <%
+	  prev_num_options = num_options
+  %>      case BlockSink::${block_io.doc_name}:
 % endfor
 % if prev_num_options != 0:
-        return ${prev_num_options};
+	  return ${prev_num_options};
 % endif
-      default:
-        return 2;
-    }
+	default:
+	  return 2;
   }
+}
 
-  /**
-   * For a given output pin, selects a given block output to use for that pin
-   * via the pin multiplexer.
-   *
-   * @param output_pin The output pin to pinmux.
-   * @param option The option to select for that pin. This value should be
-   * less than the value returned by `output_pin_options` for the given pin.
-   *
-   * The meanings of these selections can be found in the documentation:
-   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
-   */
-  bool output_pin_select(OutputPin output_pin, uint8_t option) {
-    if (option >= output_pin_options(output_pin)) {
-      Debug::log("Selected option is not valid for this pin.");
-      return false;
-    }
-    uint16_t registerOffset   = static_cast<uint16_t>(output_pin);
-    registers[registerOffset] = (1 << option);
-    return true;
-  }
+/**
+ * A handle to a sink configuration register. This can be used to select
+ * the source of the handle's associated sink.
+ */
+template <typename SinkEnum>
+struct Sink {
+  CHERI::Capability<volatile uint8_t> reg;
```

Comment:
Out of curiosity, how do you plan to handle this in the RTOS version of the driver? Currently I think you'd have to get a capability to the entire Pinmux range and then modify that to only the range of the specific sink selector register, and so while the compartmentalisation would be nicer, auditing would still show access to the entire Pinmux MMIO region even if only a small subset is being used.